Changes to additional SD-JWT VC requirements

joelposti opened the pull request:
Changes to additional SD-JWT VC requirements.

The pull request initially proposed solving https://github.com/openid/OpenID4VC-HAIP/issues/164 by adding `nbf` claim requirements. The discussion and the commits added into this pull request steered this pull request towards being a more general changeset for the additional SD-JWT VC requirements.

selfissued reviewed (changes requested):
We should not do this, for the reasons described at https://github.com/openid/oid4vc-haip/issues/164#issuecomment-2638339935 .

paulbastian commented:
I think that `nbf` actually may have privacy benefits over `iat`, as `iat` may be used for correlation, while `nbf` may be rounded up/down to UTC 00:00 for example

joelposti commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -275,7 +276,7 @@ Note: When using this profile with other cryptosuites, it is recommended to be e
 ## Validity Period of the Signature and the Claim Values
 
-`iat` and `exp` JWT claims express both the validity period of both the signature and the claims about the subject, unless there is a separate claim used to express the validity of the claims.
+`nbf` and `exp` JWT claims express both the validity period of both the signature and the claims about the subject, unless there is a separate claim used to express the validity of the claims.
```
> The `iat` claim MUST be used to evaluate the validity of a signature.

What would the validation rules be for `iat`? I do not see the connection between `iat` claim and the validity of a signature.

danielfett replied:
Discussed with @bc-pi @awoie that this sentence should be removed.

joelposti replied:
I pushed a commit where the sentence is removed. I also removed the whole section ‘Validity Period of the Signature and the Claim Values’ as the sentence was all it contained.

selfissued reviewed (changes requested):
The normal time claims in JWT-based tokens are `iat` and `exp`.  `nbf is only useful if the token is future dated - meaning that it is not valid when issued.  This is an odd corner case that we should discourage, at best.

andprian commented:
@joelposti, to answer to your comment https://github.com/openid/oid4vc-haip/issues/164#issuecomment-2668189142, you can refer to the ETSI TS 119 472-1 V0.0.6 that was made available for public review a few weeks ago: https://docbox.etsi.org/esi/Open/Latest_Drafts/ETSI%20DRAFT%20TS_119_472-1v0.0.6-public.pdf . Indeed, what is mandated in ETSI are `nbf` and `exp` . If we stick to the definitions in the RFC 7519 this makes sense:

- The "nbf" (not before) claim identifies the time before which the JWT MUST NOT be accepted for processing.
- The "iat" (issued at) claim identifies the time at which the JWT was issued. This claim can be used to determine the age of the JWT.

It is clear that `iat` is only an indication while `nbf` is a reference date for validity processing. Plus, Paul's argument is another good reason to use `nbf`.

c2bo commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -219,6 +219,7 @@ This profile defines the following additional requirements for IETF SD-JWT VCs a
 |:--- |:--- |:--- |
 |iss |MUST |[@!RFC7519], Section 4.1.1 |
 |iat |MUST |[@!RFC7519], Section 4.1.6 |
+| nbf | SHOULD (at the discretion of the issuer) | [@!RFC7519], Section 4.1.5 |
```
```suggestion
| nbf | SHOULD (at the discretion of the issuer) | [@!RFC7519], Section 4.1.5 |
```

bc-pi replied:
the whole table

joelposti replied:
I pushed a commit where the table, and the bullet points above and below it are removed.

In the same commit a new bullet point ‘Claims containing time-related information, such as issuance or expiration dates, MUST be either individually randomized within an appropriate time window (e.g., within the last 24 hours), or rounded (e.g., to the start of the day), to avoid unintended correlation factors.’ is added.

danielfett commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -219,6 +219,7 @@ This profile defines the following additional requirements for IETF SD-JWT VCs a
 |:--- |:--- |:--- |
 |iss |MUST |[@!RFC7519], Section 4.1.1 |
 |iat |MUST |[@!RFC7519], Section 4.1.6 |
+| nbf | SHOULD (at the discretion of the issuer) | [@!RFC7519], Section 4.1.5 |
```
Issue: See this issue in SD-JWT VC: https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/321

jogu replied:
Conclusion of above SD-JWT VC issue seems to be that nothing will change in sd-jwt vc, so we're back to considering what to do in HAIP.

joelposti commented:
I pushed three new commits according to the discussion in this pull request. I also rebased the branch of this pull request.

martijnharing commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,19 +233,8 @@ Requirements for both the Wallet and the Verifier:
-
-* The Issuer MUST NOT make any of the JWT Claims in the table above to be selectively disclosable, so that they are always present in the SD-JWT-VC presented by the Holder.
-* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The Wallet and the verifier  MUST support both mechanisms.
+* Claims containing time-related information, such as issuance or expiration dates, MUST be either individually randomized within an appropriate time window (e.g., within the last 24 hours), or rounded (e.g., to the start of the day), to avoid unintended correlation factors.
```
Making this a MUST is problematic for credentials where the precision is actually necessary for security reasons. As an example if you have a credential that determines access to a building, you want to have the ability to give multiple credentials the exact same access conditions.

Sakurann replied:
WG discussion: remove this clause from this PR to unblock it. open a new issue on this topic and discuss there.

Sakurann replied:
```suggestion
```

@joelposti would be great if you could commit the suggestion

Sakurann commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,20 +233,8 @@ Requirements for both the Wallet and the Verifier:
-* The Issuer MUST NOT make any of the JWT Claims in the table above to be selectively disclosable, so that they are always present in the SD-JWT-VC presented by the Holder.
-* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The Wallet and the verifier  MUST support both mechanisms.
-* The `iss` claim MUST be an HTTPS URL.
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The Wallet and the verifier MUST support both mechanisms.
```
```suggestion
* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The Wallet and the Verifier MUST support both mechanisms.
```

joelposti replied:
Thank you for the suggestions! I applied Brian's suggestion that built upon Kristina's. I leave this thread open for a while for visibility.

Sakurann commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,20 +233,9 @@ Requirements for both the Wallet and the Verifier:
+* Claims containing time-related information, such as issuance or expiration dates, MUST be either individually randomized within an appropriate time window (e.g., within the last 24 hours), or rounded (e.g., to the start of the day), to avoid unintended correlation factors.
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT-VC. The Wallet and the verifier MUST support both mechanisms.
+* The `iss` claim, if present, MUST be an HTTPS URL.
 * The `vct` JWT claim as defined in [@!I-D.ietf-oauth-sd-jwt-vc].
```
i am ok removing this bullet point

joelposti replied:
Great! I went ahead and removed the bullet point. I leave this thread open for a while so that others have a chance to comment on it.

jogu commented:
@joelposti It would be good if you could update both the title and the contents of the initial comment please - they don't seem aligned with the actual PR content anymore!

joelposti commented:
> @joelposti It would be good if you could update both the title and the contents of the initial comment please - they don't seem aligned with the actual PR content anymore!

Sure thing. I updated the initial comment and changed the title to ‘Changes to additional SD-JWT VC requirements’

bc-pi commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,21 +233,8 @@ Requirements for both the Wallet and the Verifier:
-* The `vct` JWT claim as defined in [@!I-D.ietf-oauth-sd-jwt-vc].
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Wallet and the Verifier MUST support both mechanisms.
+* The `iss` claim, if present, MUST be an HTTPS URL.
 * The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` sub claim.
```
```suggestion
* The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` member.
```

joelposti replied:
I applied your suggestion. I leave this thread open for a while for visibility.

jogu commented:
> There is currently a conflict with the `main` branch. The conflict is that the `cnf` row in the claim table, that we are removing in this pull request, has changed from
> 
> ```
> | cnf | MUST |	[@!RFC7800]|
> ```
> 
> to
> 
> ```
> | cnf | MUST if the corresponding Credential Configuration requires cryptographic holder binding | [@!RFC7800]|
> ```
> 
> Shall we go ahead with removing the table including the above addition to the `cnf` table row or shall we incorporate the changes in the `cnf` table row from `main` into the `cnf` bullet point which at the moment in this pull request is formulated like so:
> 
> ```
> * The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` member.
> ```

My personal take is that adding that 'if the corresponding Credential Configuration requires cryptographic holder binding' clarification into the `cnf`  clause you mention is a good way forward.

awoie commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,22 +233,9 @@ Requirements for both the Wallet and the Verifier:
-* The `iss` claim MUST be an HTTPS URL.
-* The `vct` JWT claim as defined in [@!I-D.ietf-oauth-sd-jwt-vc].
-* The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` sub claim.
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Wallet and the Verifier MUST support both mechanisms.
```
```suggestion
* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Wallet and the Verifier MUST support both mechanisms.
```
This seems redundant but not a hill I would die on.

jogu replied:
I'd tend to agree, I think in particular that MUST is not covered elsewhere.

I am not sure if the current language means "issuer can choose to use neither exp nor status" or if it means "issuer has to use at least one of exp & status"? That might be best handled in a new issue though.

jogu replied:
Transferred this into an issue: https://github.com/openid/OpenID4VC-HAIP/issues/256

awoie reviewed (approved):
Approving but it would be nice if @jogu 's and my last suggestion could be applied before merging.

bc-pi commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,22 +233,9 @@ Requirements for both the Wallet and the Verifier:
-* The `vct` JWT claim as defined in [@!I-D.ietf-oauth-sd-jwt-vc].
-* The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` sub claim.
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Wallet and the Verifier MUST support both mechanisms.
+* The `iss` claim, if present, MUST be an HTTPS URL.
```
This and https://openid.github.io/OpenID4VC-HAIP/openid4vc-high-assurance-interoperability-profile-wg-draft.html#section-7.1 basically, i think, say that the x509 mechanism is required but one could also support the JWT VC Issuer Metadata mechanism at the same time https://datatracker.ietf.org/doc/html/draft-ietf-oauth-sd-jwt-vc-10#name-issuer-signature-mechanisms but constrains it from being something else/more than that.

jogu replied:
Ah, thanks!

That seems kind of pointless to me given issuer metadata only seems able to hold keys, which would seem redundant if x509 based keys are required to be available and everyone is required to be able to validate via x5c - not sure if I'm missing something.

If `iss` is present and is a https url, is it optional or required for an `/.well-known/jwt-vc-issuer` based on that url to actually exist?

charsleysa replied:
> If `iss` is present and is a https url, is it optional or required for an `/.well-known/jwt-vc-issuer` based on that url to actually exist?

From my understanding, it is not required, but HAIP could make it required. 

However, as web based key resolution was removed from HAIP, there wouldn't really be any benefit to adding such a requirement.

Web based key resolution is supposed to make it simpler for Verifiers to verify the signature as there doesn't need to be pre-sharing of keys/certificates, they simply need to trust the Issuer url.

joelposti commented on openid4vc-high-assurance-interoperability-profile-1_0.md:
```diff
@@ -233,22 +233,9 @@ Requirements for both the Wallet and the Verifier:
-* The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` sub claim.
+* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Wallet and the Verifier MUST support both mechanisms.
+* The `iss` claim, if present, MUST be an HTTPS URL.
+* The `cnf` claim [@!RFC7800] MUST conform to the definition given in [@!I-D.ietf-oauth-sd-jwt-vc]. Implementations conforming to this profile MUST include the JSON Web Key [@!RFC7517] in the `jwk` member.
```
Thanks @jogu! I applied your suggestion as it helps resolving the conflict with the `main` branch.

jogu commented:
@joelposti I think we're likely good to merge this now, but it's got conflicts from one of the other recently merged PRs - sorry - could you try and resolve them please?
